File: frontend/src/components/benchmark/content/floyd_warshall/implementations/floyd_warshall.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class FloydWarshall
{
    private Dictionary<int, HashSet<int>> nodes = new Dictionary<int, HashSet<int>>();
    private Dictionary<int, Dictionary<int, int>> dist = new Dictionary<int, Dictionary<int, int>>();
    private Dictionary<int, Dictionary<int, int>> nextNode = new Dictionary<int, Dictionary<int, int>>();

    public void AddEdge(int u, int v, int weight)
    {
        if (!nodes.ContainsKey(u)) nodes[u] = new HashSet<int>();
        if (!nodes.ContainsKey(v)) nodes[v] = new HashSet<int>();
        nodes[u].Add(v);
        nodes[v].Add(u);

        if (!dist.ContainsKey(u)) dist[u] = new Dictionary<int, int>();
        if (!dist.ContainsKey(v)) dist[v] = new Dictionary<int, int>();
        dist[u][v] = weight;
        dist[v][u] = weight;

        if (!nextNode.ContainsKey(u)) nextNode[u] = new Dictionary<int, int>();
        if (!nextNode.ContainsKey(v)) nextNode[v] = new Dictionary<int, int>();
        nextNode[u][v] = v;
        nextNode[v][u] = u;
    }

    public void AddDirectedEdge(int u, int v, int weight)
    {
        if (!nodes.ContainsKey(u)) nodes[u] = new HashSet<int>();
        if (!nodes.ContainsKey(v)) nodes[v] = new HashSet<int>();
        nodes[u].Add(v);

        if (!dist.ContainsKey(u)) dist[u] = new Dictionary<int, int>();
        if (!dist.ContainsKey(v)) dist[v] = new Dictionary<int, int>();
        dist[u][v] = weight;

        if (!nextNode.ContainsKey(u)) nextNode[u] = new Dictionary<int, int>();
        if (!nextNode.ContainsKey(v)) nextNode[v] = new Dictionary<int, int>();
        nextNode[u][v] = v;
    }

    public void BuildFromEdges(List<(int, int, int)> edges, bool directed = false)
    {
        foreach (var (u, v, w) in edges)
        {
            if (directed)
                AddDirectedEdge(u, v, w);
            else
                AddEdge(u, v, w);
        }
    }

    private void Initialize()
    {
        var allNodes = GetAllNodes();

        foreach (int u in allNodes)
        {
            if (!dist.ContainsKey(u)) dist[u] = new Dictionary<int, int>();
            if (!nextNode.ContainsKey(u)) nextNode[u] = new Dictionary<int, int>();

            foreach (int v in allNodes)
            {
                if (u == v)
                {
                    dist[u][v] = 0;
                }
                else if (!dist[u].ContainsKey(v))
                {
                    dist[u][v] = int.MaxValue;
                }
            }
        }
    }

    private List<int> GetAllNodes()
    {
        return nodes.Keys.ToList();
    }

    public void AllPairsShortestPaths()
    {
        Initialize();
        var allNodes = GetAllNodes();

        foreach (int k in allNodes)
        {
            foreach (int i in allNodes)
            {
                foreach (int j in allNodes)
                {
                    if (dist[i][k] != int.MaxValue && dist[k][j] != int.MaxValue &&
                        dist[i][k] + dist[k][j] < dist[i][j])
                    {
                        dist[i][j] = dist[i][k] + dist[k][j];
                        nextNode[i][j] = nextNode[i][k];
                    }
                }
            }
        }
    }

    public int ShortestPath(int u, int v)
    {
        if (!dist.ContainsKey(u) || !dist[u].ContainsKey(v))
            return int.MaxValue;
        return dist[u][v];
    }

    public List<int> GetPath(int u, int v)
    {
        if (!nextNode.ContainsKey(u) || !nextNode[u].ContainsKey(v))
            return new List<int>();

        var path = new List<int>();
        path.Add(u);

        while (u != v)
        {
            u = nextNode[u][v];
            path.Add(u);
        }

        return path;
    }

    public Dictionary<int, Dictionary<int, int>> GetAllDistances()
    {
        return dist;
    }

    public bool HasNegativeCycle()
    {
        var allNodes = GetAllNodes();

        foreach (int i in allNodes)
        {
            if (dist[i][i] < 0)
                return true;
        }
        return false;
    }
}

File: frontend/src/components/benchmark/content/floyd_warshall/implementations/floyd_warshall.rs
use std::collections::{HashMap, HashSet};

pub struct FloydWarshall {
    nodes: HashMap<i32, HashSet<i32>>,
    dist: HashMap<i32, HashMap<i32, i32>>,
    next_node: HashMap<i32, HashMap<i32, i32>>,
}

impl FloydWarshall {
    pub fn new() -> Self {
        FloydWarshall {
            nodes: HashMap::new(),
            dist: HashMap::new(),
            next_node: HashMap::new(),
        }
    }

    pub fn add_edge(&mut self, u: i32, v: i32, weight: i32) {
        self.nodes.entry(u).or_insert_with(HashSet::new).insert(v);
        self.nodes.entry(v).or_insert_with(HashSet::new).insert(u);

        self.dist.entry(u).or_insert_with(HashMap::new).insert(v, weight);
        self.dist.entry(v).or_insert_with(HashMap::new).insert(u, weight);

        self.next_node.entry(u).or_insert_with(HashMap::new).insert(v, v);
        self.next_node.entry(v).or_insert_with(HashMap::new).insert(u, u);
    }

    pub fn add_directed_edge(&mut self, u: i32, v: i32, weight: i32) {
        self.nodes.entry(u).or_insert_with(HashSet::new).insert(v);
        self.nodes.entry(v).or_insert_with(HashSet::new);

        self.dist.entry(u).or_insert_with(HashMap::new).insert(v, weight);

        self.next_node.entry(u).or_insert_with(HashMap::new).insert(v, v);
    }

    pub fn build_from_edges(&mut self, edges: &[(i32, i32, i32)], directed: bool) {
        for &(u, v, weight) in edges {
            if directed {
                self.add_directed_edge(u, v, weight);
            } else {
                self.add_edge(u, v, weight);
            }
        }
    }

    fn get_all_nodes(&self) -> Vec<i32> {
        self.nodes.keys().cloned().collect()
    }

    fn initialize(&mut self) {
        let all_nodes = self.get_all_nodes();

        for &u in &all_nodes {
            self.dist.entry(u).or_insert_with(HashMap::new);
            self.next_node.entry(u).or_insert_with(HashMap::new);

            for &v in &all_nodes {
                if u == v {
                    self.dist.get_mut(&u).unwrap().insert(v, 0);
                } else if !self.dist.get(&u).unwrap().contains_key(&v) {
                    self.dist.get_mut(&u).unwrap().insert(v, i32::MAX);
                }
            }
        }
    }

    pub fn all_pairs_shortest_paths(&mut self) {
        self.initialize();
        let all_nodes = self.get_all_nodes();

        for &k in &all_nodes {
            for &i in &all_nodes {
                for &j in &all_nodes {
                    let dist_ik = *self.dist.get(&i).unwrap().get(&k).unwrap();
                    let dist_kj = *self.dist.get(&k).unwrap().get(&j).unwrap();
                    let dist_ij = *self.dist.get(&i).unwrap().get(&j).unwrap();

                    if dist_ik != i32::MAX && dist_kj != i32::MAX && dist_ik + dist_kj < dist_ij {
                        self.dist.get_mut(&i).unwrap().insert(j, dist_ik + dist_kj);
                        let next = *self.next_node.get(&i).unwrap().get(&k).unwrap();
                        self.next_node.get_mut(&i).unwrap().insert(j, next);
                    }
                }
            }
        }
    }

    pub fn shortest_path(&self, u: i32, v: i32) -> i32 {
        if let Some(dist_u) = self.dist.get(&u) {
            if let Some(&d) = dist_u.get(&v) {
                return d;
            }
        }
        i32::MAX
    }

    pub fn get_path(&self, u: i32, v: i32) -> Vec<i32> {
        if let Some(next_u) = self.next_node.get(&u) {
            if !next_u.contains_key(&v) {
                return Vec::new();
            }
        } else {
            return Vec::new();
        }

        let mut path = vec![u];
        let mut current = u;

        while current != v {
            current = *self.next_node.get(&current).unwrap().get(&v).unwrap();
            path.push(current);
        }

        path
    }

    pub fn get_all_distances(&self) -> &HashMap<i32, HashMap<i32, i32>> {
        &self.dist
    }

    pub fn has_negative_cycle(&self) -> bool {
        let all_nodes = self.get_all_nodes();

        for &i in &all_nodes {
            if let Some(dist_i) = self.dist.get(&i) {
                if let Some(&d) = dist_i.get(&i) {
                    if d < 0 {
                        return true;
                    }
                }
            }
        }
        false
    }
}